crates/ra_hir/src/code_model.rs

```diff
@@ -255,6 +255,15 @@ impl StructField {
     }
 }
 
+impl HasVisibility for StructField {
+    fn visibility(&self, db: &impl HirDatabase) -> ResolvedVisibility {
+        let struct_field_id: hir_def::StructFieldId = (*self).into();
+        let visibility = db.visibility(struct_field_id.into());
+        let parent_id: hir_def::VariantId = self.parent.into();
+        visibility.resolve(db, &parent_id.resolver(db))
+    }
```

Comment:
One thing I don't like about this approach is that computing visibilites for all fields of the struct will repeatedly call `child_source` method on the struct itself, which seems wasteful. For `hir_def`, I've tried to mostly stick to design which avoids such needlessly duplicate queries, by returning an `ArenaMap` for all children instead (see child_sources or field_tys as examples). I am not sure if that's a good idea, but I'd like to try to stick with this approach, as it's more "data oriented" (we express shape of the data in queries, instead of hidhling it behind abstractions). 

But this maybe a moot point if we eagarly compute visibility and store it inside `XData`

Reply:
Yeah

Reply:
Ah, this is `code_model`! Yeah, for code_model repeated queries should not be a problem. As code_model is used mostly for rendering things to the user, it should not be on a hot path.